### PR TITLE
fix #28: Promise and async/await refactor for sync logic

### LINKED ISSUE #28: refactor all callback in Promise

### PR DESCRIPTION
**Pull request by frankobe**

### CODE REVIEW

File: src/sync/index.js [hunk 1]
```diff
@@ -58,316 +58,346 @@ function sequentialLoop(iterations, process, exit){
...
+  var initialSync = sequentialLoop(Math.ceil((currentBlockChainHeight-startBlock)/batchSize), function(loop){
+    var endBlock = Math.min(startBlock + batchSize - 1, currentBlockChainHeight);
+    var syncTopic = false, syncCOracle = false, syncDOracle = false, syncVote = false, syncOracleResult = false, syncFinalResult = false;
+
+    // sync TopicCreated
+    var syncTopicCreatedPromise = new Promise(async (resolve) => {
+      let result;
+      try {
+        result = await qclient.searchLogs(startBlock, endBlock, Contracts.EventFactory.address, [Contracts.EventFactory.TopicCreated], Contracts, removeHexPrefix);
+        console.log('searchlog TopicCreated')
```
Comment (dwalintukan) [inline src/sync/index.js]: is this missing a resolve() here?

File: src/sync/index.js [hunk 2]
```diff
@@ -58,316 +58,346 @@ function sequentialLoop(iterations, process, exit){
...
+      console.log(`${startBlock} - ${endBlock}: Retrieved ${result.length} entries from TopicCreated`);
+      var createTopicPromises = [];
+
+      _.forEach(result, (event, index) => {
+        let blockNum = event.blockNumber;
+        let txid = event.transactionHash;
+        _.forEachRight(event.log, (rawLog) => {
+          if(rawLog['_eventName'] === 'TopicCreated'){
+            var insertTopicDB = new Promise(async (resolve) =>{
+              var topic = new Topic(blockNum, txid, rawLog).translate();
```
Comment (dwalintukan) [inline src/sync/index.js]: would it make more sense from an efficiency standpoint to declare these `var` outside the loop for reuse inside it?

File: src/sync/index.js [hunk 3]
```diff
@@ -58,316 +58,346 @@ function sequentialLoop(iterations, process, exit){
...
+    var syncPromises = [];
+    var updatePromises = [];
+    syncPromises.push(syncTopicCreatedPromise);
+    syncPromises.push(syncCentralizedOracleCreatedPromise);
+    syncPromises.push(syncDecentralizedOracleCreatedPromise);
+    syncPromises.push(syncOracleResultVotedPromise);
+    updatePromises.push(syncOracleResultSetPromise);
+    updatePromises.push(syncFinalResultSetPromise);
+
+
```
Comment (dwalintukan) [inline src/sync/index.js]: 👾

File: src/sync/index.js (hunk 1 above)
Comment (frankobe) [reply src/sync/index.js]: @dwalintukan nope. in the error-free flow, `resolve()` is called when db modification is doen

File: src/sync/index.js (hunk 2 above)
Comment (frankobe) [reply src/sync/index.js]: @dwalintukan I think it is fine. Only a few `var` is wasted and can be handled by gc

File: src/sync/index.js (hunk 3 above)
Comment (frankobe) [reply src/sync/index.js]: done

**Pull request merged**